### PR TITLE
Course regionalization

### PR DESCRIPTION
**Pull request by mhgoldman**

### CODE REVIEW

File: app/models/course.rb [hunk 1]
```diff
@@ -4,7 +4,17 @@ class Course < ActiveRecord::Base
   has_many :students
   has_many :instructors
 
+  has_many :course_regions
+
```
Comment (eliotsykes) [inline app/models/course.rb]: Consider adding the inverse of the `through` relationship you added in `Region` as this looks like it'll be useful elsewhere `has_many :regions, through: :course_regions`

File: app/views/courses/index.html.erb [hunk 2]
```diff
@@ -12,14 +12,18 @@
...
+          <% end %>
         </tr>
       </thead>
       <tbody>
         <% @courses.each do |course| %>
           <tr>
             <td><%= course.name %></td>
-            <td><%= course.region_name %></td>
+            <% @regions.each do |region| %>
+            <td class="text-center"><%= course.available_region_names.include?(region.name) ? glyphicon_for(:check) : glyphicon_for(:unchecked) %></td>
```
Comment (eliotsykes) [inline app/views/courses/index.html.erb]: `course.available_region_names.include?(region.name)` may be shortened to `course.regions.include?(region)` if the `through` assocation is added.

File: app/models/course_region.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class CourseRegion < ActiveRecord::Base
+  belongs_to :course
+  belongs_to :region
+
```
Comment (eliotsykes) [inline app/models/course_region.rb]: Should `CourseRegion` have an association with `Participant`?

File: app/models/participant.rb [hunk 4]
```diff
@@ -1,21 +1,44 @@
...
+
+  def course_region_name=(name)
+    if name.present? && event.course
+      @course_region_name = name
+      self.course_region = event.course.course_regions.joins(:region).
+        where("lower(regions.name) = ?", name.downcase).first
+    end
+  end
+
+  def validate_course_region_set
```
Comment (eliotsykes) [inline app/models/participant.rb]: Can `validate_course_region_set` be replaced with a `presence: { message: "..." }` validator?

Comment (eliotsykes) [inline app/models/participant.rb]: Ah I see the message varies! Not that I could find official docs, but the `message:` option appears to accept a Proc for this kind of situation: http://stackoverflow.com/a/6073458/67834

Comment (eliotsykes) [inline app/models/participant.rb]: On closer review I don't think the Proc will help here, as it doesn't accept any params so we can't analyze the model attribute value in the proc.

File: app/models/event.rb [hunk 5]
```diff
@@ -91,30 +91,36 @@ def local_end_at
...
     participants.each do |participant|
-      participant.mark_for_destruction unless emails.include?(participant.email)
+      if participant_items.none? {|item| item[:email].downcase == participant.try(:email).try(:downcase) }
+        participant.mark_for_destruction
+      end      
+    end
+
+    # Initialize any participants who've been added
+    participant_items.each do |participant_item| 
+      participant = participants.detect {|p| p.email == participant_item[:email]}
```
Comment (eliotsykes) [inline app/models/event.rb]: @mhgoldman - just realized forgot to handle initializing **new** participants with this change.

```ruby
email = participant_item[:email]
participant = participants.detect {|p| p.email == email} || participants.build(email: email)
```

Comment (mhgoldman) [inline app/models/event.rb]: Uh, good point. Thanks Eliot.

**Pull request merged**